### PR TITLE
kvstore: Add mutex Lock and Unlock metrics

### PR DESCRIPTION
**Pull request by odinuge**
Add metrics lock and unlock metrics in order to make it easier to understand how much time they take, and if there potentially is a lot of contention around these locks. Its also interesting as these result in potentially multiple etcd requests in case there is contention, and waits until it successfully has grabbed the lock.

This metric does not include the time it takes to potentially acquire a new session, or the time taken to grab the local kvstore lock for the given key.

---

Please ensure your pull request adheres to the following guidelines:

- [ ] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [ ] All code is covered by unit and/or runtime tests where feasible.
- [ ] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [ ] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [ ] Provide a title or release-note blurb suitable for the release notes.
- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [ ] Thanks for contributing!

Fixes: #issue-number

```release-note
Added Lock and Unlock metric for kvstore locks
```

---

```bash
# operator: With this PR and https://github.com/cilium/cilium/pull/36014#pullrequestreview-2445051298
$ kubectl get --raw /api/v1/namespaces/kube-system/pods/http:$(kubectl get pods -n kube-system -l io.cilium/app=operator -ojsonpath='{.items[0].metadata.name}'):9963/proxy/metrics |egrep -a "(kvstore|etcd|version)" |grep "cilium_operator_kvstore_operations_duration_seconds_count"
cilium_operator_kvstore_operations_duration_seconds_count{action="AcquireLease",kind="set",outcome="success",scope="lease"} 2
cilium_operator_kvstore_operations_duration_seconds_count{action="ListPrefix",kind="read",outcome="success",scope="identities/v1"} 1649
cilium_operator_kvstore_operations_duration_seconds_count{action="ListPrefixLocked",kind="read",outcome="success",scope="identities/v1"} 2356
cilium_operator_kvstore_operations_duration_seconds_count{action="Lock",kind="set",outcome="success",scope="cilium/.init"} 395
cilium_operator_kvstore_operations_duration_seconds_count{action="Lock",kind="set",outcome="success",scope="identities/v1"} 2356
cilium_operator_kvstore_operations_duration_seconds_count{action="Unlock",kind="delete",outcome="success",scope="cilium/.init"} 395
cilium_operator_kvstore_operations_duration_seconds_count{action="Unlock",kind="delete",outcome="success",scope="identities/v1"} 2356
cilium_operator_kvstore_operations_duration_seconds_count{action="Update",kind="set",outcome="success",scope="cilium/.hear"} 197
cilium_operator_kvstore_operations_duration_seconds_count{action="Update",kind="set",outcome="success",scope="default/cilium"} 2

# agent: With this PR
$ k get --raw /api/v1/namespaces/kube-system/pods/http:$(kubectl get pods -n kube-system -l k8s-app=cilium -ojsonpath='{.items[0].metadata.name}'):9090/proxy/metrics |grep "cilium_kvstore_operations_duration_seconds_count"
cilium_kvstore_operations_duration_seconds_count{action="AcquireLease",kind="set",outcome="success",scope="lease"} 2
cilium_kvstore_operations_duration_seconds_count{action="CreateOnly",kind="set",outcome="success",scope="identities/v1"} 21
cilium_kvstore_operations_duration_seconds_count{action="CreateOnlyLocked",kind="set",outcome="success",scope="identities/v1"} 1
cilium_kvstore_operations_duration_seconds_count{action="Delete",kind="delete",outcome="success",scope="identities/v1"} 1
cilium_kvstore_operations_duration_seconds_count{action="Delete",kind="delete",outcome="success",scope="ip/v1"} 1
cilium_kvstore_operations_duration_seconds_count{action="Get",kind="read",outcome="success",scope="identities/v1"} 22
cilium_kvstore_operations_duration_seconds_count{action="Get",kind="read",outcome="success",scope="ip/v1"} 1
cilium_kvstore_operations_duration_seconds_count{action="Get",kind="read",outcome="success",scope="nodes/v1"} 38555
cilium_kvstore_operations_duration_seconds_count{action="ListPrefix",kind="read",outcome="success",scope="identities/v1"} 1
cilium_kvstore_operations_duration_seconds_count{action="ListPrefixLocked",kind="read",outcome="success",scope="identities/v1"} 1
cilium_kvstore_operations_duration_seconds_count{action="Lock",kind="set",outcome="success",scope="cilium/.init"} 63
cilium_kvstore_operations_duration_seconds_count{action="Lock",kind="set",outcome="success",scope="identities/v1"} 1
cilium_kvstore_operations_duration_seconds_count{action="Unlock",kind="delete",outcome="success",scope="cilium/.init"} 63
cilium_kvstore_operations_duration_seconds_count{action="Unlock",kind="delete",outcome="success",scope="identities/v1"} 1
cilium_kvstore_operations_duration_seconds_count{action="Update",kind="set",outcome="success",scope="identities/v1"} 1
cilium_kvstore_operations_duration_seconds_count{action="Update",kind="set",outcome="success",scope="ip/v1"} 1
cilium_kvstore_operations_duration_seconds_count{action="Update",kind="set",outcome="success",scope="nodes/v1"} 1
cilium_kvstore_operations_duration_seconds_count{action="UpdateIfLocked",kind="set",outcome="success",scope="identities/v1"} 1

### CODE REVIEW

File: pkg/kvstore/etcd.go [hunk 1]
```diff
@@ -723,6 +727,9 @@ func (e *etcdClient) LockPath(ctx context.Context, path string) (KVLocker, error
 		return nil, Hint(err)
 	}
 
+	defer func(duration *spanstat.SpanStat) {
+		increaseMetric(path, metricSet, "Lock", duration.EndError(err).Total(), err)
```
Comment (odinuge) [inline pkg/kvstore/etcd.go:731]: Was thinking about appending/prepending a `Mutex` to get `LockMutex` but feel like both works equally.

**Review by giorio94 — APPROVED** [review]
Thanks!

File: pkg/kvstore/etcd.go (hunk 1 above)
Comment (giorio94) [reply pkg/kvstore/etcd.go:731]: I agree. The current naming makes sense to me.

**Pull request merged**